packages/babel-parser/src/util/expression-scope.js

```diff
@@ -0,0 +1,183 @@
+  }
+  recordDeclarationError(pos: number, message: string) {
+    this.declarationErrorPos = pos;
+    this.declarationErrorMessage = message;
+  }
+  hasDeclarationError(): boolean {
+    return "declarationErrorPos" in this;
+  }
+}
+
+export default class ExpressionScopeHandler {
+  stack: Array<ExpressionScope> = [new ExpressionScope()];
+  declare raise: raiseFunction;
+  constructor(raise: raiseFunction) {
+    this.raise = raise;
+  }
+  enter(scope: ExpressionScope) {
+    this.stack.push(scope);
+  }
+
+  exit() {
+    this.stack.pop();
+  }
+
+  /**
+   * Record likely parameter initializer errors
+   *
+   * When current scope is a ParameterDeclaration, the error will be thrown immediately,
+   * otherwise it will be recorded to any ancestry MaybeArrowParameterDeclaration and
+   * MaybeAsyncArrowParameterDeclaration scope until an Expression scope is seen.
+   * @param {number} pos Error position
+   * @param {string} message Error message
+   * @memberof ExpressionScopeHandler
+   */
+  recordParameterInitializerError(pos: number, message: string): void {
+    const { stack } = this;
+    let i = stack.length - 1;
+    let scope: ExpressionScope = stack[i];
+    while (!scope.isCertainlyParameterDeclaration()) {
+      if (scope.canBeArrowParameterDeclaration()) {
+        /*:: invariant(scope instanceof ArrowHeadParsingScope) */
+        scope.recordDeclarationError(pos, message);
+      } else {
+        /*:: invariant(scope.type == kExpression) */
+        // Type-Expression is the boundary where initializer error can populate to
+        return;
+      }
+      scope = stack[--i];
+    }
+    /* eslint-disable @babel/development-internal/dry-error-messages */
+    this.raise(pos, message);
+  }
+  /**
+   * Record likely async arrow parameter errors
+   *
+   * Errors will be recorded to any ancestry MaybeAsyncArrowParameterDeclaration
+   * scope until an Expression scope is seen.
+   * @param {number} pos
+   * @param {string} message
+   * @memberof ExpressionScopeHandler
+   */
+  recordAsyncArrowParametersError(pos: number, message: string): void {
+    const { stack } = this;
+    let i = stack.length - 1;
+    let scope: ExpressionScope = stack[i];
+    while (scope.canBeArrowParameterDeclaration()) {
+      if (scope.type === kMaybeAsyncArrowParameterDeclaration) {
+        /*:: invariant(scope instanceof ArrowHeadParsingScope) */
+        scope.recordDeclarationError(pos, message);
+      }
+      scope = stack[--i];
+    }
+  }
+
+  validateAsPattern(): void {
+    const { stack } = this;
+    const currentScope = stack[stack.length - 1];
+    if (!currentScope.canBeArrowParameterDeclaration()) return;
+    /*:: invariant(currentScope instanceof ArrowHeadParsingScope) */
+    if (currentScope.hasDeclarationError()) {
+      this.raise(
+        currentScope.declarationErrorPos,
+        /* eslint-disable @babel/development-internal/dry-error-messages */
+        currentScope.declarationErrorMessage,
+      );
+    }
```

Comment:
To avoid duplicate errors in the ancestor scopes, we can do something like this to delete the duplciates (assuming the Map implementation):
```js
let i = stack.length - 1;
const currentScope = stack[i];

for (const [pos, message] of currentScope.errors) {
  while (i >= 0 && stack[i].errors.delete(pos)) i--;

  this.raise(pos, message);
}
```